### PR TITLE
RD-9255: Support for calling a method with external arguments

### PR DESCRIPTION
**Pull request by bgaidioz**
* An implementation of `EnvironmentParameterEntry` for Truffle.
* A test that makes sure all types are supported when given as an external parameter, and properly handled internally.

### CODE REVIEW

File: raw-compiler-rql2-truffle/src/main/scala/raw/compiler/rql2/truffle/Rql2TruffleCompiler.scala [hunk 1]
```diff
@@ -71,6 +88,81 @@ class Rql2TruffleCompiler(implicit compilerContext: CompilerContext)
...
+                              EnvironmentPackageBuilder.Parameter(Rql2IntervalType(), StringConst(k))
+                          }
+                          // Build the argument node.
+                          // If argument is mandatory, then according to RQL2's language definition, the argument is not named.
+                          // However, if the argument os optional, the name is added.
+                          FunAppArg(e, if (mandatory.contains(k)) None else Some(k))
+                      }
+                      Right(Rql2Program(methods, Some(FunApp(IdnExp(IdnUse(decl)), programArgs.toVector))))
+                    case None =>
+                      // TODO what error should we report here? Or is it a bug?
```
Comment (bgaidioz) [inline raw-compiler-rql2-truffle/src/main/scala/raw/compiler/rql2/truffle/Rql2TruffleCompiler.scala:151]: Would it be a bug?

File: raw-compiler-rql2-truffle/src/main/scala/raw/compiler/rql2/truffle/Rql2TruffleCompiler.scala [hunk 2]
```diff
@@ -71,6 +88,81 @@ class Rql2TruffleCompiler(implicit compilerContext: CompilerContext)
...
+                      throw new CompilerException("not supported")
+                  }
+                case None =>
+                  // TODO what does it mean to have None parameters? Or is it a bug?
+                  throw new CompilerException("not supported")
```
Comment (bgaidioz) [inline raw-compiler-rql2-truffle/src/main/scala/raw/compiler/rql2/truffle/Rql2TruffleCompiler.scala:156]: Would it be a bug?

File: raw-compiler-rql2-truffle/src/main/scala/raw/compiler/rql2/truffle/Rql2TruffleCompiler.scala [hunk 3]
```diff
@@ -71,6 +88,81 @@ class Rql2TruffleCompiler(implicit compilerContext: CompilerContext)
...
+              }
+            case None =>
+              // TODO Shouldn't we add the decl name?
+              throw new CompilerException("declaration not found")
```
Comment (bgaidioz) [inline raw-compiler-rql2-truffle/src/main/scala/raw/compiler/rql2/truffle/Rql2TruffleCompiler.scala:160]: Don't we need to add the decl name?

File: raw-compiler-rql2-truffle/src/main/scala/raw/compiler/rql2/truffle/Rql2TruffleCompiler.scala [hunk 4]
```diff
@@ -71,6 +88,81 @@ class Rql2TruffleCompiler(implicit compilerContext: CompilerContext)
...
+            // executable.
+            throw new CompilerException("no expression found")
+          }
+
+          val e = me.get
+          Right(Rql2Program(methods, Some(e)))
+        case Some(decl) => tree.description.expDecls.get(decl) match {
+            case Some(descriptions) =>
+              // TODO what if there are several?
+              assert(descriptions.size == 1)
```
Comment (bgaidioz) [inline raw-compiler-rql2-truffle/src/main/scala/raw/compiler/rql2/truffle/Rql2TruffleCompiler.scala:116]: What if there are several? (We support that I think.)

File: raw-compiler-rql2/src/test/scala/raw/compiler/rql2/tests/regressions/RD9255Test.scala [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  )
+  test("three parameters")(_ =>
+    runAsRepose(
+      declarations,
+      "three_param_func",
+      Vector(("x", ParamString("U")), ("y", ParamInt(2)), ("z", ParamString("?")))
+    ) should evalTo("\"U2?\"")
+  )
+  test("two parameters")(_ =>
+    runAsRepose(declarations, "three_param_func", Vector(("x", ParamString("U")), ("y", ParamInt(2)))) should evalTo(
```
Comment (bgaidioz) [inline raw-compiler-rql2/src/test/scala/raw/compiler/rql2/tests/regressions/RD9255Test.scala:116]: rename as `run/call with param`

File: raw-runtime-rql2-truffle/src/main/java/raw/runtime/truffle/ast/expressions/builtin/environment_package/EnvironmentParameterNode.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+@ImportStatic(value = TypeGuards.class)
+@NodeInfo(shortName = "Environment.Parameter")
+@NodeChild(value = "index")
+@NodeField(name = "paramType", type = Rql2Type.class)
+public abstract class EnvironmentParameterNode extends ExpressionNode {
+
+    protected abstract Rql2Type getParamType();
+
+    @Specialization(guards = {"isByteKind(getParamType())"})
+    protected byte getByte(String index) {
```
Comment (bgaidioz) [inline raw-runtime-rql2-truffle/src/main/java/raw/runtime/truffle/ast/expressions/builtin/environment_package/EnvironmentParameterNode.java:43]: call it 'key'.

**Review by miguelbranco80 — APPROVED** [review]

**Pull request merged**